### PR TITLE
Add tests for whitespace stripping. Simplified semantics of the same

### PR DESCRIPTION
**Pull request by maghoff**
Before this change, the tests produced the following results:

For the input `<div>    </div>` (four spaces in a row), no spaces were stripped. The output was identical to the input.

For the input `<div>\n\r\n  </div>`, the following output was generated: `<div><imgsrc="image.png"> </div>`. Notice that the whitespace between `img` and `src` has been stripped entirely away, changing the semantical meaning of the HTML.

After this change, the same tests give the following output, respectively: `<div> </div>` (one space) and `<div>  </div>`. This mechanism is easier to understand, and less error prone: Any sequence of whitespace characters is replaced by one whitespace character.

Whitespace stripping still has problems. Whitespace should not be altered within `<pre>` and `<script>` blocks, for example. Inside `<script>`, replacing `\n` with ` ` will make line comments extend to the next lines, commenting out too much. However, I believe this change makes the whitespace stripping that's already present simpler and easier to understand.

### CODE REVIEW

**Comment by houd1ni** [comment]
feel free to reopen https://github.com/olado/doT/issues/191 due to maintain issues with this fork

**Comment by epoberezkin** [comment]
This change probably can be merged into 2.0.0, it is not backward compatible.

It is indeed a minor inconvenience that newline without spaces gets completely stripped changing HTML, but it is not a very common situation because there is almost always some indentation in HTML, so it is a bit theoretic...

**Comment by maghoff** [comment]
Notice that my example includes indentation in the form of tabs.

Also, this pull request comes as a result of actual problems I have experienecd. It is a bit grating to have my personal experience be classified as "theoretical".

**Comment by epoberezkin** [comment]
@maghoff sorry, I can be missing something. I thought indentation would create whitespace. I will review.

File: test/testdoT.js [hunk 1]
```diff
@@ -61,4 +61,15 @@ describe('doT', function(){
 		});
 	});
 
+	describe('whitespace stripping', function() {
+		function assert_match(actual, expected) {
+			assert(actual.match(expected), JSON.stringify(actual) + " did not match " + expected);
+		}
+		it('should replace stretches of whitespace with one whitespace', function() {
+			assert_match(doT.template('<div>    </div>')(), /^<div>\s<\/div>$/);
```
Comment (fearphage) [inline test/testdoT.js]: Why didn't you use `assert.equal`? You know the value exactly. No need to regex match it.

File: doT.js [hunk 2]
```diff
@@ -91,8 +91,7 @@
 		var cse = c.append ? startend.append : startend.split, needhtmlencode, sid = 0, indv,
 			str  = (c.use || c.define) ? resolveDefs(c, tmpl, def || {}) : tmpl;
 
-		str = ("var out='" + (c.strip ? str.replace(/(^|\r|\n)\t* +| +\t*(\r|\n|$)/g," ")
-					.replace(/\r|\n|\t|\/\*[\s\S]*?\*\//g,""): str)
+		str = ("var out='" + (c.strip ? str.replace(/\s+/g," ") : str)
```
Comment (fearphage) [inline doT.js]: You might as well trim the start and end as well.

File: test/testdoT.js (hunk 1 above)
Comment (maghoff) [inline test/testdoT.js]: It is an effort not to overspecify; the current code will accept any valid whitespace, as long as it is just one character. That is the functional requirement.

I could accept the exact string match as well, because this relaxation is not worth any confusion :)

File: doT.js (hunk 2 above)
Comment (maghoff) [inline doT.js]: Sure.

File: test/testdoT.js (hunk 1 above)
Comment (fearphage) [inline test/testdoT.js]: I understand the way the functionality works, but tests should *always* be as simple and explicit as possible. Both tests should be updated. In my opinion, the regex gives the illusion that there is some uncertainty/room for variation when that is not the case.

**Comment by maghoff** [comment]
It's been a long time since I gave up on DoT, and I am not going to put any effort into getting this PR accepted, sorry. I am now unsubscribed from this PR.

**Comment by epoberezkin** [comment]
/\s+/g would match spaces in JavaScript string literals used inside template, so cannot be used.
